waiter/src/waiter/handler.clj

```diff
@@ -1282,3 +1282,96 @@
         request))
     (catch Throwable th
       (utils/exception->response th request))))
+
+
+(defn get-service-id-from-instance-id [instance-id]
+  (nth (str/split instance-id #"\.") 0))
+
+(defn- execute-signal
+  "Helper function to send signals to instances of a service.
+   An instance needs to be approved for killing by peers before an actual kill attempt is made.
+   The function stops and returns true when a successful kill is made.
+   Else, it terminates after we have exhausted all candidate instances to kill or when a kill attempt returns a non-truthy value."
+  ;; PEER-ACK UNUSED
+  ;; dont need threadpool
+  [notify-instance-killed-fn peers-acknowledged-eject-requests-fn allowed-to-manage-service?-fn scheduler populate-maintainer-chan! timeout-config
+   instance-id correlation-id thread-pool response-chan]
+  (let [{:keys [eject-backoff-base-time-ms inter-kill-request-wait-time-ms max-eject-time-ms]} timeout-config]
+    (cid/with-correlation-id
+      correlation-id
+      (async/go
+        (try
+          ;; DONT KNOW IF WE NEED REASON AND RESULT MAP
+          (let [request-id (utils/unique-identifier) ; new unique identifier for this reservation request
+                reason-map-fn (fn [] {:cid correlation-id :reason :kill-instance :request-id request-id :time (t/now)})
+                result-map-fn (fn [status] {:cid correlation-id :request-id request-id :status status})]
+            (log/info "Attempting to kill" instance-id)
+              ;; MAKE OUR OWN TIMER FOR INSTANCES?
+              ;; (metrics/service-timer instance-id "kill-instance")
+              ;;loop [exclude-ids-set #{}]
+                ;;let [instance (service/get-rand-inst populate-maintainer-chan! service-id (reason-map-fn) exclude-ids-set inter-kill-request-wait-time-ms)]
+                (let [service-id (get-service-id-from-instance-id instance-id)]
+                    (def instance {:id instance-id :service-id service-id})
+                      (do
+                        ;; CHANGED INSTANCE TO INSTANCE-ID
+                        (log/info "sending sigkill to instance " instance-id service-id)
+                        ;; (counters/inc! (metrics/service-counter service-id "scaling" "scale-down" "attempt"))
+                        (scheduler/track-kill-candidate! instance-id :prepare-to-kill eject-backoff-base-time-ms)
+                        (let [{:keys [killed?] :as kill-result}
+                              (-> (au/execute
+                                    (fn send-signal-to-instance []
+                                      (scheduler/signal-instance scheduler instance :sigkill))
```

Comment:
We would like the `signal-instance` function to be non-blocking. And we can remove the threadpool dependency.